python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,446 @@
+    """
+    Time class is for both Date transforms and Timestamp transforms.
+    """
+
+    _TIME_SATISFIED_ORDER = dict(year=3, month=2, day=1, hour=0)
+    _VALID_TIME_GRANULARITY = {
+        DateType: {"year", "month", "day"},
+        TimestampType: {"year", "month", "day", "hour"},
+        TimestamptzType: {"year", "month", "day", "hour"},
+    }
+    _INSTANCES: dict = {DateType: {}, TimestampType: {}, TimestamptzType: {}}
+
+    def __new__(cls, source_type: Type, name: str):
+        if cls._INSTANCES.get(source_type, {}).get(name) is None:
+            if name not in Time._VALID_TIME_GRANULARITY.get(source_type, {}):
+                raise ValueError(f"Cannot partition type: {source_type} by {name}")
+            cls._INSTANCES[source_type][name] = super(Time, cls).__new__(cls)
+        return cls._INSTANCES[source_type][name]
+
+    def __init__(self, source_type: Type, name: str):
+        super().__init__(
+            name,
+            f"transforms.{name}(source_type={repr(source_type)})",
+            getattr(transform_util, f"human_{name}"),
+        )
+        self._type = source_type
+        self._name = name
+
+        self._diff_func = getattr(transform_util, f"diff_{self._name}")
+        if self._name == "day" and self._type == DateType:
+            self._apply = lambda v: v
+        elif self._type == DateType:
+            self._apply = lambda v: self._diff_func(
+                datetime.utcfromtimestamp(v * 86400)
+            )
+        else:
+            self._apply = lambda v: self._diff_func(
+                datetime.utcfromtimestamp(v / 1000000)
+            )
+
+        self._result_type = DateType if self._name == "day" else IntegerType
+
+    def apply(self, value: int) -> int:
+        return self._apply(value)
+
+    def can_transform(self, target: Type) -> bool:
+        if self._type == DateType:
+            return target == DateType
+        else:  # self._type is either TimestampType or TimestamptzType
+            return target == TimestampType or target == TimestamptzType
+
+    def result_type(self, source_type: Type) -> Type:
+        return self._result_type
+
+    def preserves_order(self) -> bool:
+        return True
+
+    def satisfies_order_of(self, other: Transform) -> bool:
+        if self == other:
+            return True
+
+        if isinstance(other, Time):
+            return (
+                Time._TIME_SATISFIED_ORDER[self._name]
+                <= Time._TIME_SATISFIED_ORDER[other._name]
+            )
+
+        return False
+
+    def dedup_name(self) -> str:
+        return "time"
+
+
+class Identity(Transform):
+    _HUMAN_STRING_MAP = {
+        DateType: lambda v: transform_util.human_day(v),
+        TimeType: lambda v: transform_util.human_time(v),
+        TimestampType: lambda v: transform_util.human_timestamp(v),
+        TimestamptzType: lambda v: transform_util.human_timestamptz(v),
+        BinaryType: lambda v: transform_util.base64encode(v),
```

Comment:
In general, I think this PR would benefit from using the same approach as `types` and just create a few instances of transforms named `DateIdentityTransform` or similar. I think this is trying to be too clever by looking up functions by source type when it would be easy to read and maintain if you had something like this:

```python
DateIdentityTransform = IdentityTransform(DateType, tranform_util.human_day)
TimeIdentityTransform = IdentityTransform(TimeType, transform_util.human_time)
FixedIdentityTransform = IdentityTransform(FixedType, transform_util.base64encode)
...
```

Reply:
SG

Reply:
I created some singletons with special to_human string and leave others to be created at runtime.